extensions/elytron-security-oauth2/runtime/src/main/java/io/quarkus/elytron/security/oauth2/runtime/oauth/ElytronOAuth2CallerPrincipal.java

```diff
@@ -0,0 +1,36 @@
+package io.quarkus.elytron.security.oauth2.runtime.oauth;
+
+import java.security.Principal;
+
+import org.wildfly.security.authz.Attributes;
+
+/**
+ * An implementation of ElytronOAuth2CallerPrincipal that builds on the Elytron attributes
+ */
+public class ElytronOAuth2CallerPrincipal implements Principal {
+    private Attributes claims;
+    private String customPrincipalName;
+
+    public ElytronOAuth2CallerPrincipal(final String customPrincipalName, final Attributes claims) {
+        this.claims = claims;
+        this.customPrincipalName = customPrincipalName;
+    }
+
+    public ElytronOAuth2CallerPrincipal(final Attributes claims) {
+        this("username", claims);
+    }
+
+    public Attributes getClaims() {
```

Comment:
I'm not sure the user should deal with `Attributes` if the access to the other introspection response claims is needed, may be have another method returning a Map

Reply:
This is copied from the ElytronJwtCallerPrincipal from the smallrye-jwt extention. I'm agree that a Map should be a better implementation.
Maybe this can be done later in an effort to consolidate the three implementations (token, jwt, keyclock). We can then use a single CallerPrincipal that we build with a name (so the logic to find the name in the claims is moved to the creator of the principal) and a map of attributes that are computed from the claims.

Reply:
`ElytronJwtCallerPrincipal` may get refactored or even go if the smallrye-jwt extension is refactored to use the factory mechanism; a public `Attributes` getter is redundant in that context because `JsonWebToken` interface which `ElytronJwtCallerPrincipal` implements provides an access to all the claims, so accessing the claims via `Attributes` is not what the smallrye-jwt users would be doing IMHO.
If we have a `Map` it will not block us from adding `Attributes` at the later stage if all agree, but if have `Attributes` now we won't be able to remove them later and the JAX-RS client option (to avoid the wildlfly library indirection) won't work either if we choose to simplify later...

Reply:
I don't know why but Elytron `Attributes` allow a list of string a entry value and not only a String.
As the introspection endpoint return a JSON and that multiple attribute with the same name is not allowed in JSON we can map attributes by using the first value of each key.
But we will loose the capability of expressing multi-value key (for example to pre-extract a list from a value with comma delimited). Don't know if it's useful ...

Reply:
I think that if a given claim value in the introspection response is say `1,2,3` then it has to be returned as is as only the application code which will know what that means, JSON array will be returned by any proper endpoint if it is meant to be multi-valued. And we won't lose this option :-), it will be there for us to use if really needed (may be we will use `Attributes` internally to pass it to smallrye-jwt to convert them to `JsonWebToken` or for something else). 
To be honest I don't mind if no `Map` will be offered just yet either. May be a `JSONB` would better be returned, just feel that hiding the `Attributes` for now will give us more options later... Not a  big deal though

Reply:
I use a `Map<String, List<String>>` instead of the Attributes object.
As you said, an attribute can be a JSON array, that's why the value class of the Map is a List of String.

Reply:
@loicmathieu thanks. It should be an Object, not ideal, but List<String> won't work if the claim is a complex object in which case a value would become a Map as well. Likewise it can be an array of objects. As I said I don't mind Map not being provided just yet either but if you agree it makes sense to have from the start then please tweak it further a bit. Just having a Map<String, Object> being returned would work :-) 

The user would have to cast based on the application-level understanding of what the given claim really is...I guess most users won't need it but some might

Reply:
done